cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java

```diff
@@ -60,26 +63,76 @@
+
+  /**
+   * Holds the total number of requests that have been executed by this handler that should count toward max allowed.
+   * If a request is not valid, or does not involve user code execution, it should not be increment.
+   * If set to {@link TaskWorkerHttpHandlerInternal#FORCE_RESTART}, pod will get restarted as soon as the ongoing
+   * request finishes.
+   */
+  private final AtomicInteger totalRequestsWithUserCodeExecution = new AtomicInteger(0);
+
   private final String metadataServiceEndpoint;
 
   public TaskWorkerHttpHandlerInternal(CConfiguration cConf, Consumer<String> stopper) {
+    int containerKillAfterExecution = cConf.getInt(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION, 0);
     this.runnableTaskLauncher = new RunnableTaskLauncher(cConf);
     this.metadataServiceEndpoint = cConf.get(Constants.TaskWorker.METADATA_SERVICE_END_POINT);
     this.stopper = (terminate, className) -> {
-      if (cConf.getBoolean(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION) && terminate) {
-        if (className != null) {
-          stopper.accept(className);
-        }
+      if (totalRequestsWithUserCodeExecution.get() != FORCE_RESTART &&
+        (!terminate || className == null || containerKillAfterExecution <= 0)) {
+        // No need to restart.
+        totalRequestsWithUserCodeExecution.decrementAndGet();
+        inflightRequests.set(0);
+        return;
+      }
+
+      if (totalRequestsWithUserCodeExecution.get() >= containerKillAfterExecution) {
+        stopper.accept(className);
       } else {
         inflightRequests.set(0);
       }
     };
+
+    enablePeriodicRestart(cConf, stopper);
+  }
+
+  /**
+   * If there is no ongoing request, worker pod gets restarted after a random duration from the following
+   * range is select. Otherwise, worker pod gets restarted once the ongoing request finishes.
+   * range = [Duration - DURATION_FRACTION * Duration, Duration + DURATION_FRACTION * Duration]
+   * Reason: by randomizing the duration, it is guaranteed that pods do not get restarted at the same time.
+   */
+  private void enablePeriodicRestart(CConfiguration cConf, Consumer<String> stopper) {
+    int duration = cConf.getInt(Constants.TaskWorker.CONTAINER_KILL_AFTER_DURATION, 0);
+    int lowerBound = (int) (duration - duration * DURATION_FRACTION);
+    int upperBound = (int) (duration + duration * DURATION_FRACTION);
+    if (lowerBound > 0) {
+      int waitTime = (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
+      Executors.newSingleThreadScheduledExecutor().schedule(
+        () -> {
+          if (inflightRequests.compareAndSet(0, 1)) {
+            // No ongoing request. restart immediately.
+            stopper.accept("");
+            return;
+          }
+          // there is an ongoing request. Restart once the request finishes.
+          totalRequestsWithUserCodeExecution.set(FORCE_RESTART);
+          this.stopper.accept(true, "");
```

Comment:
This is confusing to have `stopper` referring to either a `Consumer` or `BiConsumer`.

Reply:
Also I have some hard time following the logic. How calling accept(true, "") will delay the restart until the inflight request finished?? Can you simplify the logic?